### PR TITLE
some refactoring

### PR DESCRIPTION
**Pull request by ikitiki**
* all acronyms must consist of capital letters
* move some functions out of logicalbackup struct to dbutils
* multiple minor refactorings
* split big methods into smaller ones

### CODE REVIEW

File: pkg/dbutils/dbutils.go [hunk 1]
```diff
@@ -80,3 +86,83 @@ func QuoteLiteral(str string) string {
 		return `'` + res + `'`
 	}
 }
+
+func InitPublication(conn *pgx.Conn, publicationName string) error {
+	rows, err := conn.Query("select 1 from pg_publication where pubname = $1;", publicationName)
+	if err != nil {
+		return fmt.Errorf("could not execute query: %v", err)
+	}
+
```
Comment (alexeyklyukin) [inline pkg/dbutils/dbutils.go]: defer rows.Close() (noticed we are not doing it if publication creation errors out, but we still should.

Comment (ikitiki) [reply pkg/dbutils/dbutils.go]: we do close it(https://github.com/ikitiki/logical_backup/pull/35/files#diff-87ca3ef81cadbba95ef6810c5ae9051cR97 and https://github.com/ikitiki/logical_backup/pull/35/files#diff-87ca3ef81cadbba95ef6810c5ae9051cR100)

Comment (alexeyklyukin) [reply pkg/dbutils/dbutils.go]: ah, right. But doing it in a defer avoids duplicating rows.Close() all over the function.

File: pkg/dbutils/dbutils.go [hunk 2]
```diff
@@ -80,3 +86,83 @@ func QuoteLiteral(str string) string {
 		return `'` + res + `'`
 	}
 }
+
+func InitPublication(conn *pgx.Conn, publicationName string) error {
```
Comment (alexeyklyukin) [inline pkg/dbutils/dbutils.go]: I'd name it CreatePublicationIfNotExists, as we don't do any actions if the publication is there.

File: pkg/dbutils/dbutils.go [hunk 3]
```diff
@@ -80,3 +86,83 @@ func QuoteLiteral(str string) string {
...
+	if _, err := conn.Exec(query); err != nil {
+		return fmt.Errorf("could not create publication: %v", err)
+	}
+	rows.Close()
+	log.Printf("created missing publication: %q", query)
+
+	return nil
+}
+
+func InitSlot(conn *pgx.Conn, slotName, dbName string) (LSN, error) {
```
Comment (alexeyklyukin) [inline pkg/dbutils/dbutils.go]: How about GetLogicalSlotFlushLSN?

File: pkg/decoder/decoder.go [hunk 4]
```diff
@@ -196,7 +195,18 @@ func Parse(src []byte) (message.Message, error) {
 			Raw: make([]byte, len(src)),
 		}
 		copy(m.Raw, src)
-		//TODO
+
+		m.Relations = d.uint32()
```
Comment (alexeyklyukin) [inline pkg/decoder/decoder.go]: This field could be named relationsCount, but in fact we don't need it, as it's only useful to learn how many relationOids are there.

File: pkg/logicalbackup/logicalbackup.go [hunk 5]
```diff
@@ -172,109 +160,121 @@ func New(ctx context.Context, stopCh chan struct{}, cfg *config.Config) (*Logica
...
-		if err := os.Mkdir(cfg.ArchiveDir, os.ModePerm); err != nil {
-			return nil, fmt.Errorf("could not create archive dir: %v", err)
-		}
-	} else if err != nil {
-		return nil, fmt.Errorf("%q stat error: %v", cfg.ArchiveDir, err)
-	}
+	return nil
+}
+
+func (b *LogicalBackup) initDB() error {
```
Comment (alexeyklyukin) [inline pkg/logicalbackup/logicalbackup.go]: The naming clash with the Postgres tool is unfortunate; perhaps connectAndPrepareDatabase?

Comment (ikitiki) [reply pkg/logicalbackup/logicalbackup.go]: I'd name it `prepareDB`

File: pkg/logicalbackup/logicalbackup.go [hunk 6]
```diff
@@ -692,7 +595,7 @@ func (b *LogicalBackup) prepareTablesForPublication(conn *pgx.Conn) error {
 
 		for rows.Next() {
 			tab := tableInfo{}
-			err = rows.Scan(&tab.oid, &tab.name.Namespace, &tab.name.Name, &tab.hasPK, &tab.replicaIdentity)
+			err := rows.Scan(&tab.oid, &tab.name.Namespace, &tab.name.Name, &tab.hasPK, &tab.replicaIdentity)
```
Comment (alexeyklyukin) [inline pkg/logicalbackup/logicalbackup.go]: not sure if  idiomatic go consider this acceptable, but I'd write
```
if rows.Scan(... ) != nil {
    return
```
as we don't need to preserve err.

File: pkg/logicalbackup/logicalbackup.go [hunk 7]
```diff
@@ -749,30 +652,30 @@ func (b *LogicalBackup) prepareTablesForPublication(conn *pgx.Conn) error {
...
 
 	if err := yaml.NewDecoder(fp).Decode(&stateInfo); err != nil {
-		return dbutils.InvalidLsn, fmt.Errorf("could not decode state info yaml: %v", err)
+		return dbutils.InvalidLSN, fmt.Errorf("could not decode state info yaml: %v", err)
 	}
 
 	currentLSN, err := pgx.ParseLSN(stateInfo.CurrentLSN)
 	if err != nil {
-		return dbutils.InvalidLsn, fmt.Errorf("could not parse %q Lsn string: %v", stateInfo.CurrentLSN, err)
+		return dbutils.InvalidLSN, fmt.Errorf("could not parse %q LSN string: %v", stateInfo.CurrentLSN, err)
```
Comment (alexeyklyukin) [inline pkg/logicalbackup/logicalbackup.go]: I thought the idea was to use lowercase in the user-visible messages, but uppercase in the code (certainly this is not consistent with usage of lsn in other error messages).

File: pkg/logicalrestore/logicalrestore.go [hunk 8]
```diff
@@ -334,23 +334,23 @@ func (r *LogicalRestore) checkTableStruct() error {
 	return nil
 }
 
-func (r *LogicalRestore) getOid(tblName message.NamespacedName) (dbutils.Oid, error) {
-	history := make(map[dbutils.Oid][]logicalbackup.NameAtLsn)
+func (r *LogicalRestore) getOid(tblName message.NamespacedName) (dbutils.OID, error) {
```
Comment (alexeyklyukin) [inline pkg/logicalrestore/logicalrestore.go]: shouldn't it be getOID ?

**Comment by alexeyklyukin** [comment]
Apart from same naming issues highlighted above I have no other objections. Great job fixing all those inconsistencies (and some actual errors)!

File: pkg/logicalbackup/logicalbackup.go (hunk 7 above)
Comment (alexeyklyukin) [reply pkg/logicalbackup/logicalbackup.go]: well, then we have change them the other way around in other messages (we can do it as a separate PR not to delay merging this one).

File: pkg/logicalbackup/logicalbackup.go (hunk 6 above)
Comment (ikitiki) [reply pkg/logicalbackup/logicalbackup.go]: the problem here is that you're shadowing the `err` which is defined prior that line

Comment (ikitiki) [reply pkg/logicalbackup/logicalbackup.go]: ah. now I've figured out that shadowing is intended here

File: pkg/decoder/decoder.go [hunk 9]
```diff
@@ -78,12 +78,11 @@ func (d *decoder) columns() []message.Column {
 	size := int(d.uint16())
 	data := make([]message.Column, size)
 	for i := 0; i < size; i++ {
-		data[i] = message.Column{
-			IsKey:   d.bool(),
-			Name:    d.string(),
-			TypeOID: d.Oid(),
-			Mode:    d.int32(),
-		}
+		data[i] = message.Column{}
```
[the remark this answers is not in the document]
Comment (ikitiki) [reply pkg/decoder/decoder.go]: I'd keep it as it is. as it's basically copy-paste from [that](https://github.com/kyleconroy/pgoutput) project

**Pull request merged**